storage/pin/pin.go

```diff
@@ -0,0 +1,400 @@
+// down the merkle tree and pin all the chunks that are encountered on the
+// way. It pins both data chunk and tree chunks. The pre-requisite is that
+// the file should be present in the local database. This function is called
+// from two places 1) Just after the file is uploaded 2) anytime after
+// uploading the file using the pin command. This function can pin both
+// encrypted and non-encrypted files.
+func (p *API) PinFiles(rootHash string, isRaw bool, credentials string) error {
+	addr, err := hex.DecodeString(rootHash)
+	if err != nil {
+		log.Error("Error decoding root hash", "rootHash", rootHash, "err", err)
+		return err
+	}
+
+	hasChunk, err := p.db.Has(context.TODO(), chunk.Address(p.removeDecryptionKeyFromChunkHash(addr)))
+	if !hasChunk {
+		log.Error("Could not pin hash. File not uploaded", "Hash", rootHash)
+		return err
+	}
+
+	// Walk the root hash and pin all the chunks
+	walkerFunction := func(ref storage.Reference) error {
+		chunkAddr := p.removeDecryptionKeyFromChunkHash(ref)
+		err := p.db.Set(context.TODO(), chunk.ModeSetPin, chunkAddr)
+		if err != nil {
+			log.Error("Could not pin chunk. Address "+"Address", hex.EncodeToString(chunkAddr))
+			return err
+		} else {
+			log.Trace("Pinning chunk", "Address", hex.EncodeToString(chunkAddr))
+		}
+		return nil
+	}
+	p.walkChunksFromRootHash(rootHash, isRaw, credentials, walkerFunction)
+
+	// Check if the root hash is already pinned and add it to the fileInfo struct
+	if val, ok := p.pinnedFiles[rootHash]; !ok {
+		// Get the file size from the root chunk first 8 bytes
+		hashFunc := storage.MakeHashFunc(storage.DefaultHash)
+		isEncrypted := len(addr) > hashFunc().Size()
+		getter := storage.NewHasherStore(p.db, hashFunc, isEncrypted, chunk.NewTag(0, "show-chunks-tag", 0))
+		chunkData, err := getter.Get(context.TODO(), addr)
+		if err != nil {
+			log.Error("Error getting chunk data from localstore.", "Address", hex.EncodeToString(addr))
+			return nil
+		}
+		fileSize := chunkData.Size()
+
+		// Get the pin counter from the pinIndex
+		pinCounter, err := p.getPinCounterOfChunk(chunk.Address(p.removeDecryptionKeyFromChunkHash(addr)))
+		if err != nil {
+			log.Error("Error getting pin counter of root hash.", "rootHash", rootHash, "err", err)
+			return nil
+		}
+
+		// Store it in the fileinfo data structure
+		// this is pushed to state DB when Close() is called
+		p.pinnedFiles[rootHash] = FileInfo{
+			isRaw:      isRaw,
+			fileSize:   fileSize,
+			pinCounter: pinCounter,
+		}
+	} else {
+		// Get the pin counter from the pinIndex
+		pinCounter, err := p.getPinCounterOfChunk(chunk.Address(p.removeDecryptionKeyFromChunkHash(addr)))
+		if err != nil {
+			log.Error("Error getting pin counter of root hash.", "rootHash", rootHash, "err", err)
+			return nil
+		}
+
+		val.pinCounter = pinCounter
+		p.pinnedFiles[rootHash] = val
+	}
+
+	// Store the pinned files in state DB
+	err = p.state.Put("pin-files", p.pinnedFiles)
```

Comment:
What if UnpinFiles and PinFiles method is called concurrently, both storing (Put) pinnedFiles map? There can be a data race where this map is accessed and modified concurrently and also stored resulting in data corruption.

Reply:
Yeah.. possible.. added mutex locks.

Reply:
Thanks.